### PR TITLE
[Merged by Bors] - feat(Algebra/Star/NonUnitalSubsemiring): Add NonUnitalStarSubsemiring

### PR DESCRIPTION
**Pull request by mans0954**
Definitions and basic properties of non-unital star subsemirings

---

This file heavily copies `Mathlib.Algebra.Star.NonUnitalSubalgebra`, so I'm happy to assign the copyright to @j-loreaux if that's the appropriate thing to do.

So far I've mostly only developed what's required for #6595 - I thought I'd check that this was the right approach before going further.

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Algebra/Star/NonUnitalSubsemiring.lean [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+
+-/
+
+universe v w w'
+
+variable {A : Type v} {B : Type w} {C : Type w'}
+
+/-- A non-unital star subset is a subset which is closed under the `star`
+operation. -/
+structure StarSubset (A : Type v) [Star A] : Type v where
```
Comment (j-loreaux) [inline Mathlib/Algebra/Star/NonUnitalSubsemiring.lean:29]: I'm dubious that we want this. Do you have applications in mind for sets that are closed *only* under the `star` operation?

File: Mathlib/Algebra/Star/NonUnitalSubsemiring.lean [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+  /-- The underlying set of a `SubStar`. -/
+  carrier : Set A
+  /-- The `carrier` of a `StarSubset` is closed under the `star` operation. -/
+  star_mem' : ∀ {a : A} (_ha : a ∈ carrier), star a ∈ carrier
+
+/-- A sub star magma is a subset of a magma which is closed under the `star`-/
+structure SubStarmagma (M : Type v) [Mul M] [Star M] extends Subsemigroup M, StarSubset M : Type v
```
Comment (j-loreaux) [inline Mathlib/Algebra/Star/NonUnitalSubsemiring.lean:36]: At the very least, the capitalization is wrong, but also, I think we should match naming with `Subsemigroup`. If we want to switch the naming of both to `Magma`, then fine, but that should be a separate PR.
```suggestion
/-- A sub star magma is a subset of a magma which is closed under the `star`-/
structure SubStarSemigroup (M : Type v) [Mul M] [Star M] extends Subsemigroup M, StarSubset M : Type v
```

File: Mathlib/Algebra/Star/NonUnitalSubsemiring.lean [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+    extends NonUnitalSubsemiring R, StarSubset R : Type v
+
+/-- Reinterpret a `NonUnitalStarSubsemiring` as a `NonUnitalSubsemiring`. -/
+add_decl_doc NonUnitalStarSubsemiring.toNonUnitalSubsemiring
+/-- Reinterpret a `NonUnitalStarSubsemiring` as a `StarSubset`. -/
+add_decl_doc NonUnitalStarSubsemiring.toStarSubset
+
+/-- A (unital) star subsemiring is a non-associative ring which is closed under the `star`
+operation. -/
+structure StarSubsemiring (R : Type v) [NonAssocSemiring R] [Star R]
```
Comment (j-loreaux) [inline Mathlib/Algebra/Star/NonUnitalSubsemiring.lean:55]: It's weird that this is in the `NonUnitalSubsemiring` file. Perhaps you can move the unital stuff out of here into a separate file (and possibly even a separate PR).

File: Mathlib/Algebra/Star/NonUnitalSubsemiring.lean [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+/-- A sub star magma is a subset of a magma which is closed under the `star`-/
+structure SubStarmagma (M : Type v) [Mul M] [Star M] extends Subsemigroup M, StarSubset M : Type v
+
+/-- Reinterpret a `SubStarmagma` as a `Subsemigroup`. -/
+add_decl_doc SubStarmagma.toSubsemigroup
+/-- Reinterpret a `SubStarmagma` as a `StarSubset`. -/
+add_decl_doc SubStarmagma.toStarSubset
+
+/-- A non-unital star subsemiring is a non-unital non-associative ring which is closed under the
+`star` operation. -/
```
Comment (j-loreaux) [inline Mathlib/Algebra/Star/NonUnitalSubsemiring.lean:44]: associativity, or the lack thereof, mostly comes from the ambient ring.
```suggestion
/-- A non-unital star subsemiring is a non-unital subsemiring which also is closed under the
`star` operation. -/
```

File: Mathlib/Algebra/Star/NonUnitalSubsemiring.lean [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+  mul_mem {s} := s.mul_mem'
+  zero_mem {s} := s.zero_mem'
+
+instance instStarMemClass : StarMemClass (NonUnitalStarSubsemiring R) R where
+  star_mem {s} := s.star_mem'
+
+theorem mem_carrier {s : NonUnitalStarSubsemiring R} {x : R} : x ∈ s.carrier ↔ x ∈ s :=
+  Iff.rfl
+
+/-- Copy of a non-unital star subalgebra with a new `carrier` equal to the old one.
```
Comment (j-loreaux) [inline Mathlib/Algebra/Star/NonUnitalSubsemiring.lean:85]:
```suggestion
/-- Copy of a non-unital star subsemiring with a new `carrier` equal to the old one.
```

File: Mathlib/Algebra/Star/NonUnitalSubsemiring.lean [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+theorem copy_eq (S : NonUnitalStarSubsemiring R) (s : Set R) (hs : s = ↑S) : S.copy s hs = S :=
+  SetLike.coe_injective hs
+
+section Center
+
+variable (R)
+
+/-- The center of a semiring `R` is the set of elements that commute and associate with everything
+in `R` -/
+def center (R) [NonUnitalNonAssocSemiring R][StarRing R] : NonUnitalStarSubsemiring R where
```
Comment (j-loreaux) [inline Mathlib/Algebra/Star/NonUnitalSubsemiring.lean:109]:
```suggestion
/-- The center of a non-unital non-associative semiring `R` is the set of elements that
commute and associate with everything in `R`, here realized as non-unital star 
subsemiring. -/
def center (R) [NonUnitalNonAssocSemiring R] [StarRing R] : NonUnitalStarSubsemiring R where
```

File: Mathlib/Algebra/Star/NonUnitalSubsemiring.lean [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+end StarSubset
+
+section SubStarmagma
+
+variable (A) [Mul A] [StarMul A]
+
+namespace SubStarmagma
+
+/-- The center of `A` is the set of elements that commute and associate
+with everything in `A` -/
```
Comment (j-loreaux) [inline Mathlib/Algebra/Star/NonUnitalSubsemiring.lean:226]:
```suggestion
/-- The center of magma `A` is the set of elements that commute and associate
with everything in `A`, here realized as a `SubStarSemigroup`. -/
```

**Review by j-loreaux — COMMENTED** [review]
I think we don't want `StarSubset`, unless you have an explicit application in mind. Also, it would be nice if the unital versions of all these were placed in a separate file.

**Comment by j-loreaux** [comment]
no need to include me in the copyright.

File: Mathlib/Algebra/Star/NonUnitalSubsemiring.lean (hunk 2 above)
Comment (mans0954) [reply Mathlib/Algebra/Star/NonUnitalSubsemiring.lean:36]: Possibly I should have used  `NonAssocSemigroup` in #7060 instead of `Magma` for consistency with `(NonUnital)NonAssoc(Semi)ring`?

File: Mathlib/Algebra/Star/NonUnitalSubsemiring.lean (hunk 1 above)
Comment (mans0954) [reply Mathlib/Algebra/Star/NonUnitalSubsemiring.lean]: No application in mind at present. Originally I was thinking there might be some efficiency in having a common base class, but I'm not convinced that's actually the case, so I have removed it now.

File: Mathlib/Algebra/Star/Subsemiring.lean [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+instance starMemClass : StarMemClass (StarSubsemiring R) R where
+  star_mem {s} := s.star_mem'
+
+instance subsemiringClass : SubsemiringClass (StarSubsemiring R) R where
+  add_mem {s} := s.add_mem'
+  mul_mem {s} := s.mul_mem'
+  zero_mem {s} := s.zero_mem'
+  one_mem {s} := s.one_mem'
+
+-- this uses the `Star` instance `s` inherits from `StarMemClass (StarSubalgebra R A) A`
```
Comment (j-loreaux) [inline Mathlib/Algebra/Star/Subsemiring.lean:46]:
```suggestion
-- this uses the `Star` instance `s` inherits from `StarMemClass (StarSubsemiring R A) A`
```

**Review by j-loreaux — APPROVED** [review]
bors d+
